### DISCUSSION TITLE
Suggestion for some references to write efficient casadi code

### DISCUSSION THREAD

**Discussion by sandeep026** [casadi/casadi#3397] [Q&A] [answered]
I am presently working on a module to transcribe and solve general optimal control problems numerically using direct methods. 
Although I have solved several optimal control problems numerically in the past, I want to write my casadi codes  as effficiently as possible. Any suggestion is highly appreciated.

For eg. , using map() instead of for loops have helped me a lot in the past while constructing the collocation/shooting constraints.

**Comment by jgillis** [comment, marked as answer]
There are some tricks and insights offered in the self-paced master class. In general,  it's hard to script a recipe for succes. Benchmarking code (e.g. via codegeneration) can help to focus on bottlenecks. As a developer, it's very easy to bump into ideas to improve casadi source code when benchmarking an application code.

**Reply by sandeep026** [reply]
Thank @jgillis .

**Comment by sandeep026** [comment]
The wiki page for onboarding is quite useful.

https://github.com/casadi/casadi/wiki/Onboarding-Guide
